### ISSUE TITLE
[question] How to compile an example using CUDA without cmake?

### ISSUE THREAD

**Issue by xiaolin-cheng** [davisking/dlib#1582]
Hi Davis,

Again thanks for your great help!

I am trying to compile an example with CUDA support not using cmake:

`g++   -std=c++11 -O3 -I/software/dlib /software/dlib/dlib/all/source.cpp -lpthread   -lX11 -ljpeg -L/usr/local/cuda/lib64 -lcublas -lcudart -DDLIB_JPEG_SUPPORT  -DDLIB_USE_BLAS -DDLIB_USE_CUDA ../dnn_metric_learning_on_images_ex.cpp`

But it gives me the error:

`tmp/cciNWpJq.o:   In function `main':     dnn_metric_learning_on_images_ex.cpp:(.text.startup+0x18a7): undefined   reference to `dlib::gpu_data::copy_to_host() const'     collect2: error: ld returned 1 exit status`

Am I missing anything?

Thank you!

**Comment by davisking** [comment]
Use cmake.

**Comment by xiaolin-cheng** [comment]
Thanks @davisking 

Any possibility to compile a single cpp file using command line g++? Am I missing any library?

**Comment by davisking** [comment]
No, use cmake to build dlib if you want cuda support.  Building cuda stuff is complicated.  You have to use nvcc instead of gcc for instance.

**Comment by DorianAppeldoorn** [comment]
Hello Davis,

I have a followup question for this thread.

The thing is that I have a large project, which is not built by CMake, that requires some functionality I would implement using dlib. I tried making a library, configuring it with CMake to use Dlib with CUDA and cuDNN functionality, but I can't get it to link to the other project. I understand that you need CMake to configure dlib with CUDA, which builds correctly, but how come I can't link it into the other project when it's statically built? All the dependencies should be inside the lib file, right? 

Would you know any way around this? Or should I really configure the whole project with CMake?

**Comment by davisking** [comment]
In general, in C++, if you build a static library out of some C++ code it
is still required that you link to whatever dependencies that static
library links to.  The compiler isn't going to pull in all downstream
libraries and jam them into the static library you built, that make static
library files unreasonably large.  On many platforms, if you use shared
libraries instead it will automatically pull in dependencies, but this is
not guaranteed.  If you get a linker error you need to link to the thing
the error is telling you to link to.

TLDR, you can certainly build dlib as a library with cmake and then build
your stuff without cmake.  Many people do this.  But you have to understand
how to setup a C++ build system.  There are many build systems and all are
configured differently.  The answer to this is not dlib specific.  Consult
the documentation for whatever build system/compiler/os/whatever you are
using.
